service/history/ndc/history_replicator_test.go

```diff
@@ -380,3 +382,1110 @@ func TestApplyEvents(t *testing.T) {
+					branchIndex int,
+					releaseFn execution.ReleaseFunc,
+					task replicationTask,
+					r *historyReplicatorImpl,
+				) error {
+					return nil
+				}
+				return fn
+			},
+			expectError: execution.ErrMissingVersionHistories,
+		},
+		"Case2: case nil with applyNonStartEventsPrepareBranchFn error": {
+			mockExecutionCacheAffordance: func(mockExecutionCache *execution.MockCache, mockExecutionContext *execution.MockContext) {
+				mockExecutionCache.EXPECT().GetOrCreateWorkflowExecution(gomock.Any(), gomock.Any(), gomock.Any()).
+					Return(mockExecutionContext, func(err error) {}, nil).Times(1)
+			},
+			mockReplicationTaskAffordance: func(mockReplicationTask *MockreplicationTask) {
+				mockReplicationTask.EXPECT().getDomainID().Return("test-domain-id").Times(1)
+				mockReplicationTask.EXPECT().getExecution().Return(&types.WorkflowExecution{
+					WorkflowID: "test-workflow-id",
+					RunID:      "test-run-id",
+				}).Times(1)
+				mockReplicationTask.EXPECT().getFirstEvent().Return(&types.HistoryEvent{
+					EventType: &workflowExecutionType,
+				}).Times(1)
+				mockReplicationTask.EXPECT().getVersion().Return(int64(1)).Times(1)
+			},
+			mockExecutionContextAffordance: func(mockExecutionContext *execution.MockContext, mockExecutionMutableState *execution.MockMutableState) {
+				mockExecutionContext.EXPECT().LoadWorkflowExecutionWithTaskVersion(gomock.Any(), gomock.Any()).
+					Return(mockExecutionMutableState, nil).Times(1)
+			},
+			mockMutableStateAffordance: func(mockExecutionMutableState *execution.MockMutableState) {
+				mockExecutionMutableState.EXPECT().GetVersionHistories().Return(&persistence.VersionHistories{
+					CurrentVersionHistoryIndex: 1,
+					Histories:                  nil,
+				}).Times(1)
+			},
+			applyNonStartEventsPrepareBranchFnAffordance: func() func(ctx ctx.Context,
+				context execution.Context,
+				mutableState execution.MutableState,
+				task replicationTask,
+				newBranchManager newBranchManagerFn,
+			) (bool, int, error) {
+				fn := func(ctx ctx.Context,
+					context execution.Context,
+					mutableState execution.MutableState,
+					task replicationTask,
+					newBranchManager newBranchManagerFn,
+				) (bool, int, error) {
+					return false, 0, fmt.Errorf("test error")
+				}
+				return fn
+			},
+			applyNonStartEventsPrepareMutableStateFnAffordance: func(mockExecutionMutableState *execution.MockMutableState) func(ctx ctx.Context,
+				context execution.Context,
+				mutableState execution.MutableState,
+				branchIndex int,
+				task replicationTask,
+				newConflictResolver newConflictResolverFn,
+			) (execution.MutableState, bool, error) {
+				fn := func(ctx ctx.Context,
+					context execution.Context,
+					mutableState execution.MutableState,
+					branchIndex int,
+					task replicationTask,
+					newConflictResolver newConflictResolverFn,
+				) (execution.MutableState, bool, error) {
+					return nil, false, fmt.Errorf("test error")
+				}
```

Comment:
You probably want to do at least some assertions here to ensure that you're passing in the right data: 

Most of the args are pretty hard to mess up, but Ints, strings, ints etc are much easier. Consider at least putting a couple of assertions here:
```
				fn := func(ctx ctx.Context,
					context execution.Context,
					mutableState execution.MutableState,
					branchIndex int,
					task replicationTask,
					newConflictResolver newConflictResolverFn,
				) (execution.MutableState, bool, error) {
                    assert.Equal(t, 123, branchIndex)
					return nil, false, fmt.Errorf("test error")
				}
				return fn
```

Reply:
resolved offline